src/entities/utils/getBoostedPoolStateWithBalancesV3.ts

```diff
@@ -0,0 +1,110 @@
+import {
+    createPublicClient,
+    erc4626Abi,
+    formatEther,
+    formatUnits,
+    http,
+} from 'viem';
+
+import { HumanAmount } from '@/data';
+import { CHAINS, VAULT_V3 } from '@/utils';
+
+import {
+    PoolStateWithUnderlyingBalances,
+    PoolStateWithUnderlyings,
+} from '../types';
+import { vaultExtensionAbi_V3 } from '@/abi';
+import { getSortedTokens } from './getSortedTokens';
+import { TokenAmount } from '../tokenAmount';
+
+export const getBoostedPoolStateWithBalancesV3 = async (
+    poolState: PoolStateWithUnderlyings,
+    chainId: number,
+    rpcUrl: string,
+): Promise<PoolStateWithUnderlyingBalances> => {
+    const totalSupplyContract = {
+        address: VAULT_V3[chainId],
+        abi: vaultExtensionAbi_V3,
+        functionName: 'totalSupply' as const,
+        args: [poolState.address] as const,
+    };
+    const getBalanceContracts = {
+        address: VAULT_V3[chainId],
+        abi: vaultExtensionAbi_V3,
+        functionName: 'getCurrentLiveBalances' as const,
+        args: [poolState.address] as const,
+    };
+
+    const publicClient = createPublicClient({
+        transport: http(rpcUrl),
+        chain: CHAINS[chainId],
+    });
+    const outputs = await publicClient.multicall({
+        contracts: [totalSupplyContract, getBalanceContracts],
+    });
+
+    if (outputs.some((output) => output.status === 'failure')) {
+        throw new Error(
+            'Error: Unable to get pool state with balances for v3 pool.',
+        );
+    }
+
+    const totalShares = outputs[0].result as bigint;
+    const balancesScale18 = outputs[1].result as bigint[];
+
+    const poolTokens = getSortedTokens(poolState.tokens, chainId);
+    const tokenBalances = poolTokens.map((token, i) =>
+        TokenAmount.fromScale18Amount(token, balancesScale18[i]),
+    );
+
+    const getUnderlyingBalancesContracts = poolState.tokens // TODO: not sorted
```

Comment:
Does this need to be sorted?

Reply:
Daniel always said we should not blindly expect the API to return sorted tokens, so we should always sort locally by index. That's how it's handled everywhere else in the code (with `getSortedTokens` helper).
I added the comment so I wouldn't forget to do it before getting it ready for review - apparently I forgot 😅

Reply:
I fixed it and refactored the code to improve readability - let me know what you think